zkevm-circuits/src/rlp_circuit.rs

```diff
@@ -0,0 +1,2368 @@
+                        meta.query_advice(length_acc, Rotation::cur()),
+                    );
+                },
+            );
+
+            // if tag_index < tag_length && tag_index > 1
+            cb.condition(
+                is_gas(meta) * tindex_lt_tlength.clone() * tindex_lt.clone(),
+                |cb| {
+                    cb.require_equal(
+                        "[gas] value_acc::next == value_acc::cur * 256 + value::next",
+                        meta.query_advice(rlp_table.value_acc, Rotation::next()),
+                        meta.query_advice(rlp_table.value_acc, Rotation::cur()) * 256.expr() +
+                            meta.query_advice(byte_value, Rotation::next()),
+                    );
+                },
+            );
+
+            // tag_index > 1
+            cb.condition(is_gas(meta) * tindex_lt.clone(), |cb| {
+                cb.require_equal(
+                    "tag::next == RlpTxTag::Gas",
+                    meta.query_advice(rlp_table.tag, Rotation::cur()),
+                    RlpTxTag::Gas.expr(),
+                );
+                cb.require_equal(
+                    "tag_index::next == tag_index - 1",
+                    meta.query_advice(rlp_table.tag_index, Rotation::next()),
+                    meta.query_advice(rlp_table.tag_index, Rotation::cur()) - 1.expr(),
+                );
+                cb.require_equal(
+                    "tag_length::next == tag_length",
+                    meta.query_advice(tag_length, Rotation::next()),
+                    meta.query_advice(tag_length, Rotation::cur()),
+                );
+            });
+
+            // tag_index == 1
+            cb.condition(is_gas(meta) * tindex_eq.clone(), |cb| {
+                cb.require_equal(
+                    "tag::next == RlpTxTag::ToPrefix",
+                    meta.query_advice(rlp_table.tag, Rotation::next()),
+                    RlpTxTag::ToPrefix.expr(),
+                );
+            });
+
+            //////////////////////////////////////////////////////////////////////////////////////
+            ///////////////////////////////// RlpTxTag::ToPrefix /////////////////////////////////
+            //////////////////////////////////////////////////////////////////////////////////////
+            cb.condition(is_to_prefix(meta), |cb| {
+                cb.require_equal(
+                    "tag_index == 1",
+                    meta.query_advice(rlp_table.tag_index, Rotation::cur()),
+                    1.expr(),
+                );
+                cb.require_equal(
+                    "tag_length == 1",
+                    meta.query_advice(tag_length, Rotation::cur()),
+                    1.expr(),
+                );
+                cb.require_equal(
+                    "value == 148",
+                    meta.query_advice(byte_value, Rotation::cur()),
+                    148.expr(),
+                );
```

Comment:
Need to handle the other case that `value = 0x80` for contract creation. Actually we can use the constraints for previous tags to handle `To` except we add one additional requirement for `value == 0x94` which is `tag_length=21`.

Reply:
Even for `callee_address` being a zero address (contract creation), the RLP encoding will be:
```
[148, 0, 0, 0, ..., 0, 0, 0]
```
As RLP encodes it as `bytes`, not as a single `value`.

Reply:
In the case of contract creation, the callee_address is zero which is often encoded as `0x80`. For example, this [tx(aztec's turbo verifier)](https://etherscan.io/getRawTx?tx=0x3e9633f1ff36b26d3bac55fb51c73fc4cc43b07080f4c877e23a1071715d6665) has `to` encoded as `0x80`.